File: python/problems/arrays_hashing/valid_anagram.py
"""Valid Anagram.

Question:
Given two strings ``s`` and ``t``, return ``True`` when ``t`` is an anagram
of ``s``. An anagram uses the same characters with the same counts.

This file includes a brute-force sorting solution and an optimized
frequency-counting solution.

How to test:
Test file: tests/python/test_valid_anagram.py
Run: PYTHONDONTWRITEBYTECODE=1 python3 -m unittest -v tests/python/test_valid_anagram.py
"""


def is_anagram_brute_force(s: str, t: str) -> bool:
    """Sort both strings and compare them.

    Time: O(n log n), because sorting n characters takes n log n time.
    Space: O(n), because sorted creates character lists.
    """
    if len(s) != len(t):
        return False

    return sorted(s) == sorted(t)


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """Count characters in one pass over each string.

        Time: O(n), because n + n is still O(n).
        Space: O(k), where k is the number of different characters stored.
        """
        if len(s) != len(t):
            return False

        count = {}

        for char in s:
            count[char] = count.get(char, 0) + 1

        for char in t:
            if count.get(char, 0) == 0:
                return False

            count[char] -= 1

        return True

File: javascript/problems/arrays_hashing/validAnagram.js
/**
 * Valid Anagram
 *
 * Question:
 * Given two strings `s` and `t`, return `true` when `t` is an anagram of `s`.
 * An anagram uses exactly the same characters with exactly the same counts.
 *
 * This file includes a brute-force sorting solution and an optimized
 * frequency-counting solution. Use the optimized version in an interview.
 *
 * How to test:
 * Test file: tests/javascript/validAnagram.test.js
 * Run: node --test tests/javascript/validAnagram.test.js
 */

/**
 * Brute force: sort both strings and compare the sorted results.
 *
 * Time: O(n log n) because sorting n characters takes n log n time.
 * We sort twice, but 2 * n log n is still O(n log n).
 * Space: O(n) because split creates character arrays.
 *
 * @param {string} s
 * @param {string} t
 * @returns {boolean}
 */
function isAnagramBruteForce(s, t) {
  if (s.length !== t.length) {
    return false;
  }

  return s.split("").sort().join("") === t.split("").sort().join("");
}

/**
 * Optimized: count each character in `s`, then subtract while reading `t`.
 *
 * Time: O(n) because we scan each string once: n + n = 2n, which is O(n).
 * Each count lookup is O(1) on average.
 * Space: O(k), where k is the number of different characters we store.
 *
 * @param {string} s
 * @param {string} t
 * @returns {boolean}
 */
function isAnagram(s, t) {
  if (s.length !== t.length) {
    return false;
  }

  const count = Object.create(null);

  for (const char of s) {
    count[char] = (count[char] || 0) + 1;
  }

  for (const char of t) {
    if (!count[char]) {
      return false;
    }

    count[char] -= 1;
  }

  return true;
}

// Keep the repository's original function name available as an alias.
const validAnagram = isAnagram;

module.exports = { isAnagram, isAnagramBruteForce, validAnagram };
